### PR TITLE
feat(autoscaling): deprecate launch configurations (under feature flag)

### LINKED ISSUE #23165: aws-autoscaling: Use launch templates instead of deprecated launch configurations
### Describe the feature

We just got an email from AWS saying launch configurations are deprecated. New accounts won't allow creating launch configurations after December 31st, 2023. No new features or instance types will be available after December 31st, 2022.

Unless given a launch template, [AutoScalingGroup](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_autoscaling.AutoScalingGroup.html) always creates a deprecated launch configuration.

### Use Case

Deploying existing CDK apps that use [AutoScalingGroup](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_autoscaling.AutoScalingGroup.html) without manually specifying a manually created launch template would fail soon. Using new instance types will be impossible without first manually creating a launch template.

### Proposed Solution

It would be great if [AutoScalingGroup](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_autoscaling.AutoScalingGroup.html) could start creating launch templates instead of launch configurations.

### Other Information

_No response_

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.52.0

### Environment details (OS name and version, etc.)

N/A

### PR DESCRIPTION
**Pull request by colifran**
This PR focuses on supporting the deprecation of launch configurations when creating an auto scaling group. All new accounts created after December 31, 2023 will no longer be able to create launch configurations. An equivalent launch template will now be created by the AutoScalingGroup construct instead of a launch configuration.

Closes #23165

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by colifran** [comment]
Still in process - do not review.

File: packages/@aws-cdk-testing/framework-integ/test/aws-ecs-patterns/test/ec2/integ.alb-ecs-service-command-entry-point.ts [hunk 1]
```diff
@@ -4,8 +4,9 @@ import * as ecs from 'aws-cdk-lib/aws-ecs';
 import * as cdk from 'aws-cdk-lib';
 import * as integ from '@aws-cdk/integ-tests-alpha';
 import * as ecsPatterns from 'aws-cdk-lib/aws-ecs-patterns';
+import { AUTOSCALING_DISABLE_LAUNCH_CONFIG } from 'aws-cdk-lib/cx-api';
 
-const app = new cdk.App();
+const app = new cdk.App({ postCliContext: { [AUTOSCALING_DISABLE_LAUNCH_CONFIG]: false } });
```
Comment (colifran) [inline packages/@aws-cdk-testing/framework-integ/test/aws-ecs-patterns/test/ec2/integ.alb-ecs-service-command-entry-point.ts:9]: This integ test is unsuccessful as it is, i.e., without any changes. The behavior is the same regardless of whether a launch configuration or a launch template is used. I'm continuing to investigate, but didn't want this to be a blocker since the issue is unrelated to the changes in this PR.

File: packages/@aws-cdk/integ-tests-alpha/lib/assertions/api-call-base.ts [hunk 2]
```diff
@@ -125,6 +125,7 @@ export abstract class ApiCallBase extends Construct implements IApiCall {
   protected expectedResult?: string;
   protected flattenResponse: string = 'false';
   protected stateMachineArn?: string;
+  protected outputPaths: string[] = [];
```
Comment (colifran) [inline packages/@aws-cdk/integ-tests-alpha/lib/assertions/api-call-base.ts:128]: Changes related to outputPaths were to fix a bug found where using getAtt and getAttString were not updating the outputPaths. This was an unrelated fix that I decided to address as part of this PR since it was impacting the aws-autoscaling-hooktargets integ test called integ.queue-hook

File: packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts [hunk 3]
```diff
@@ -1280,51 +1282,71 @@ export class AutoScalingGroup extends AutoScalingGroupBase implements
         throw new Error('Setting \'instanceType\' is required when \'launchTemplate\' and \'mixedInstancesPolicy\' is not set');
       }
 
+      Tags.of(this).add(NAME_TAG, this.node.path);
```
Comment (colifran) [inline packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts:1285]: I'd like for someone to take a second look at this. Should we only be adding this tag as part of a launch configuration? The PR this was added in mentions that this was added to support tagging of AutoScalingGroups and Security Groups so I'm thinking this isn't specific to launch configurations or launch templates, but AutoScalingGroups as a whole.

File: packages/@aws-cdk-testing/framework-integ/test/aws-autoscaling-hooktargets/test/integ.queue-hook.js.snapshot/integ-autoscalinghook-queue.template.json [hunk 4]
```diff
@@ -451,39 +451,85 @@
...
+        "Value": "integ-autoscalinghook-queue/Group/LaunchTemplate"
+       }
       ]
      }
-    ],
-    "UserData": {
-     "Fn::Base64": "#!/bin/bash"
-    }
-   },
-   "DependsOn": [
```
Comment (corymhall) [inline packages/@aws-cdk-testing/framework-integ/test/aws-autoscaling-hooktargets/test/integ.queue-hook.js.snapshot/integ-autoscalinghook-queue.template.json:476]: It's losing the dependency on the `Role`. It still has the implicit dependency on the instance profile because of the reference, but there needs to be an explicit dependency on the role.

File: packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts [hunk 5]
```diff
@@ -1280,51 +1282,71 @@ export class AutoScalingGroup extends AutoScalingGroupBase implements
...
         assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
       });
-
       this.grantPrincipal = this._role;
 
       if (props.ssmSessionPermissions) {
         this.role.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMManagedInstanceCore'));
       }
 
-      const iamProfile = new iam.CfnInstanceProfile(this, 'InstanceProfile', {
```
Comment (corymhall) [inline packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts:1302]: We need a way to keep using this with the new `LaunchTemplate`

**Review by corymhall — CHANGES_REQUESTED** [review]
@colifran looks great! Just a couple of minor comments.

File: packages/@aws-cdk-testing/framework-integ/test/aws-autoscaling-hooktargets/test/integ.queue-hook.js.snapshot/integ-autoscalinghook-queue.template.json (hunk 4 above)
Comment (colifran) [reply packages/@aws-cdk-testing/framework-integ/test/aws-autoscaling-hooktargets/test/integ.queue-hook.js.snapshot/integ-autoscalinghook-queue.template.json:476]: Good catch. I've added a launch template dependency on the role.

File: packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/ecs/integ.ec2-run-task.js.snapshot/aws-sfn-tasks-ecs-ec2-integ.template.json [hunk 6]
```diff
@@ -497,7 +497,7 @@
     ]
    }
   },
-  "Ec2ClusterDefaultAutoScalingGroupInstanceProfileDB232471": {
+  "Ec2ClusterDefaultAutoScalingGroupLaunchTemplateProfileA74899D4": {
```
[the remark this answers is not in the document]
Comment (colifran) [reply packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/ecs/integ.ec2-run-task.js.snapshot/aws-sfn-tasks-ecs-ec2-integ.template.json:500]: I came up with a solution to address this. I'm updating the integ tests to reflect this change.

File: packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts (hunk 5 above)
Comment (colifran) [reply packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts:1302]: I came up with a solution to keep using this with the new `LaunchTemplate` - this will be reflected when all changes are pushed

File: packages/aws-cdk-lib/cx-api/FEATURE_FLAGS.md [hunk 7]
```diff
@@ -1008,4 +1010,28 @@ is not viable in some productive setups.
 | 2.84.0 | `false` | `true` |
 
 
+### @aws-cdk/aws-autoscaling:disableDefaultLaunchConfigCreation
```
Comment (corymhall) [inline packages/aws-cdk-lib/cx-api/FEATURE_FLAGS.md:1013]: Since we are now creating a default launch template I think we should change this name since it implies that when enabled we won't create anything.

Maybe `defaultLaunchTemplateInsteadOfLaunchConfig`. It's kind of verbose, so if you have anything better...

File: packages/aws-cdk-lib/aws-ec2/lib/launch-template.ts [hunk 8]
```diff
@@ -786,6 +817,13 @@ export class LaunchTemplate extends Resource implements ILaunchTemplate, iam.IGr
     }
   }
 
+  /**
+   * @internal
+   */
+  public _addIamInstanceProfileArn(iamInstanceProfileArn: string) {
```
Comment (corymhall) [inline packages/aws-cdk-lib/aws-ec2/lib/launch-template.ts:823]: I feel like we should just bit the bullet and create a `InstanceProfile` L2. It would be a super simple L2 and would allow us to just pass it in through properties of the launch template

**Review by corymhall — CHANGES_REQUESTED** [review]

File: packages/aws-cdk-lib/aws-ec2/lib/launch-template.ts (hunk 8 above)
Comment (colifran) [reply packages/aws-cdk-lib/aws-ec2/lib/launch-template.ts:823]: I created an `InstanceProfile` L2 and have the functionality we want working using the L2. I need to finish testing the `InstanceProfile` L2 and then I will push these changes.

File: packages/aws-cdk-lib/aws-ec2/lib/launch-template.ts [hunk 9]
```diff
@@ -404,6 +403,20 @@ export interface LaunchTemplateProps {
...
+  /**
+   * Whether instances should have a public IP addresses associated with them.
+   *
+   * @default - Use subnet settings
+   */
+  readonly associatePublicIpAddress?: boolean;
+
+  /**
+   * The instance profile used to pass role information to EC2 instances.
+   *
```
Comment (corymhall) [inline packages/aws-cdk-lib/aws-ec2/lib/launch-template.ts:416]: Add a note here about providing either this or role (and do the same for role)

File: packages/aws-cdk-lib/aws-iam/lib/instance-profile.ts [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+/**
+ * Properties of an Instance Profile
+ */
+export interface InstanceProfileProps {
+  /**
+   * The name of the role to associate with the InstanceProfile. Only one role can
+   * be assigned to an EC2 instance at a time, and all applications on the instance
+   * share the same role and permissions.
+   */
+  readonly role: IRole;
```
Comment (corymhall) [inline packages/aws-cdk-lib/aws-iam/lib/instance-profile.ts:39]: I keep going back and forth on this, but I'm leaning towards making this optional and creating a role for you by default.

**Review by corymhall — CHANGES_REQUESTED** [review]
Looking good! Just a couple minor things

File: packages/aws-cdk-lib/aws-iam/lib/instance-profile.ts (hunk 10 above)
Comment (colifran) [reply packages/aws-cdk-lib/aws-iam/lib/instance-profile.ts:39]: This makes sense. I'll get this updated.

File: packages/aws-cdk-lib/aws-ec2/lib/launch-template.ts (hunk 9 above)
Comment (colifran) [reply packages/aws-cdk-lib/aws-ec2/lib/launch-template.ts:416]: Updated.

**Review by corymhall — APPROVED** [review]
